File: flix_backend/lib/flix_backend/data/comment.ex
defmodule FlixBackend.Data.Comment do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @derive {Phoenix.Param, key: :id}
  @derive {Jason.Encoder, only: [
    :id,
    :product_id,
    :user_id,
    :parent_id,
    :root_id,
    :content,
    :likes_count,
    :replies_count,
    :status,
    :inserted_at,
    :updated_at,
    :user,
    :is_liked
  ]}

  schema "comments" do
    field :content, :string
    field :likes_count, :integer, default: 0
    field :replies_count, :integer, default: 0
    field :status, FlixBackend.Data.CommentStatus, default: :active
    field :is_liked, :boolean, virtual: true, default: false

    belongs_to :product, FlixBackend.Data.Product, type: :binary_id
    belongs_to :user, FlixBackend.Data.User, type: :binary_id, foreign_key: :user_id, references: :uid
    belongs_to :parent, FlixBackend.Data.Comment, type: :binary_id, foreign_key: :parent_id
    belongs_to :root, FlixBackend.Data.Comment, type: :binary_id, foreign_key: :root_id

    has_many :replies, FlixBackend.Data.Comment, foreign_key: :parent_id
    has_many :likes, FlixBackend.Data.CommentLike, foreign_key: :comment_id

    timestamps()
  end

  def changeset(comment, attrs) do
    comment
    |> cast(attrs, [:product_id, :user_id, :parent_id, :root_id, :content, :status])
    |> validate_required([:product_id, :user_id, :content])
    |> validate_length(:content, min: 1, max: 1000)
  end
end

File: Flix/app/src/main/java/fin/phoenix/flix/data/Comment.kt
package fin.phoenix.flix.data

import android.annotation.SuppressLint
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize
import java.util.Date

/**
 * 评论数据类
 */
@SuppressLint("ParcelCreator")
@Parcelize
data class Comment(
    val id: String,
    @SerializedName("product_id") val productId: String,
    @SerializedName("user_id") val userId: String,
    @SerializedName("parent_id") val parentId: String? = null,
    @SerializedName("root_id") val rootId: String? = null,
    val content: String,
    @SerializedName("likes_count") var likesCount: Int = 0,
    @SerializedName("replies_count") val repliesCount: Int = 0,
    val status: CommentStatus = CommentStatus.ACTIVE,
    @SerializedName("inserted_at") val createdAt: Date,
    @SerializedName("updated_at") val updatedAt: Date,
    val user: UserAbstract? = null,
    // 客户端状态，不从服务端获取
    @SerializedName("is_liked") var isLiked: Boolean = false,
    var isHighlighted: Boolean = false
) : Parcelable

/**
 * 评论状态枚举
 */
enum class CommentStatus {
    @SerializedName("active") ACTIVE,
    @SerializedName("deleted") DELETED,
    @SerializedName("hidden") HIDDEN
}

/**
 * 评论上下文数据类，用于显示一条评论及其所在的评论树
 */
@Parcelize
data class CommentContext(
    @SerializedName("root_comment") val rootComment: Comment,
    val replies: List<Comment>,
    @SerializedName("target_comment_id") val targetCommentId: String
) : Parcelable